### PR TITLE
Added Bitcoin mining Script

### LINKED ISSUE #856: Script for BitCoin Mining in Python
# Description

It's a script to mine bitcoin but It actually won't reward any bitcoin as the difficulty level is 20 which requires huge processors for rewarding.
So, It's a fun project to understand how hashing, etc works.
I will give complete details in a pdf file to understand everything from scratch.
I would Like to work on this issue as well as part of GSSOC-21.
Thanks

## Language

- [ ] Bash
- [ ] Go
- [ ] Javascript
- [ ] Powershell
- [x] Python
- [ ] Other (Please specify)

## Checklist:

- **Details on the Script**

| name            | about                                                  | labels  | assignees                                                      |
| --------------- | ------------------------------------------------------ | ------- | -------------------------------------------------------------- |
| BitCoin_Mining | A Script for mining BitCoin| feature | @NEERAJAP2001|

### PR DESCRIPTION
**Pull request by NEERAJAP2001**
# Description

Please include a summary of the change and which issue is fixed. List any dependencies that are required for this change.

Fixes #856 

## Type of change

_Please delete options that are not relevant._

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Documentation Update

# Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by NEERAJAP2001** [comment]
@vybhav72954 Review, please!
Thanks

File: Python/BitCoin Mining/README.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+# Draw Backs:
+The script has no drawbacks but due to increase in miners , the difficultl level increases and hence we'd require best hardware as fastest wins. 
+
+# Refrences 
+https://www.investopedia.com/terms/b/bitcoin.asp
+
+https://www.youtube.com/watch?v=wTC31ZI6QM4
+
+https://www.youtube.com/watch?v=ZhnJ1bkIWWk&t=143s
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:46]: Add hyperlinks to them

File: Python/BitCoin Mining/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+ 
+
+
+
+
+
+
+
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:55]: Remove extra lines here

File: Python/BitCoin Mining/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+2. Now, Let's again assume that we have previous hash as : 'a5x208fecf9a66be9a2bc7...'
+3. Now we create text as  : text = str(block_number) + transactions + previous_hash + str(nonce)
+4. Now we pass text to SHA256 Function to generate hash
+5. Finally, that hash prefix must be equal to number of zeroes as difficulty level.
+6. And Boom!! You have mined succesfully.
+
+
+
+# Draw Backs:
+The script has no drawbacks but due to increase in miners , the difficultl level increases and hence we'd require best hardware as fastest wins.
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:39]: Lots of spelling mistakes

File: Python/BitCoin Mining/README.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Btc will be rewarded if Nonce gives us string with prefix of Zeroes the difficult number of times. sounds confusing?
+
+So, Let's take example :
+
+1. Let's assume we have difficulty level = 20 
+2. Now, Let's again assume that we have previous hash as : 'a5x208fecf9a66be9a2bc7...'
+3. Now we create text as  : text = str(block_number) + transactions + previous_hash + str(nonce)
+4. Now we pass text to SHA256 Function to generate hash
+5. Finally, that hash prefix must be equal to number of zeroes as difficulty level.
+6. And Boom!! You have mined succesfully.
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:34]: Can you use proper markdown and run your content through Grammarly? Please.

File: Python/BitCoin Mining/README.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+3. Block: Blocks are files where data pertaining to the Bitcoin network are permanently recorded
+
+4. Nonce : Miners have to guess this number which will reward them Btc , A nonce is an abbreviation for "number only used once," which is a number added to a hashed—or encrypted—block in a blockchain that, when rehashed, meets the difficulty level restrictions. 
+
+5. Difficulty Level : Number of prefix Contagiuos Zeroes 
+
+
+
+
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:22]: Remove extra lines here

File: Python/BitCoin Mining/README.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+
+# What Is Bitcoin Mining?
+Bitcoin mining is performed by high-powered computers that solve complex computational math problems; these problems are so complex that they cannot be solved by hand and are complicated enough to tax even incredibly powerful computers.
+
+
+
+
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:7]: Remove extra lines here

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Comment by NEERAJAP2001** [comment]
@HarshCasper  Done

**Comment by NEERAJAP2001** [comment]
@vybhav72954 review, please!

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+MAX_NONCE = 10000
+
+
+# Function for encoding text to a 64 bit hexadecimal value
+def SHA256(text):
+    return sha256(text.encode("ascii")).hexdigest()
+
+
+# function for guessing nonce value
+def mine(block_number, transactions, previous_hash, prefix_zeros):
```
Comment (vybhav72954) [inline Python/BitCoin Mining/BitCoin_Mining.py:21]: Add Docstrings

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+# Import the sha256 function
+from hashlib import sha256
+
+# Nonce value 
+MAX_NONCE = 10000
+
+
+# Function for encoding text to a 64 bit hexadecimal value
+def SHA256(text):
```
Comment (vybhav72954) [inline Python/BitCoin Mining/BitCoin_Mining.py:16]: Add Docstrings

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    previous_hash=input('Enter Previous has value : ')
+
+
+    # Calling mine function with all required parameters
+    new_hash = mine(5,transactions,previous_hash, difficulty)
+
+    # total time for refrence
+    total_time = str((time.time() - start))
+    print(f"end mining. Mining took: {total_time} seconds")
+    print(new_hash)
```
Comment (vybhav72954) [inline Python/BitCoin Mining/BitCoin_Mining.py:65]: Add a blank line at EOF

File: Python/BitCoin Mining/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# What Is Bitcoin Mining?
+Bitcoin mining is performed by high-powered computers that solve complex computational math problems; these problems are so complex that they cannot be solved by hand and are complicated enough to tax even compelling computers.
+
+# Terminology :
```
Comment (vybhav72954) [inline Python/BitCoin Mining/README.md:4]: Add levels of heading, every heading is H1 here, add other levels as well.

File: Python/BitCoin Mining/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
+# What Is Bitcoin Mining?
```
Comment (vybhav72954) [inline Python/BitCoin Mining/README.md:1]: After very heading there should be a blank line

File: Python/BitCoin Mining/README.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+BTC will be rewarded if Nonce gives us a string with Zeroes' prefix the difficulty number of times. Sounds confusing?
+So, Let's take an example :
+
+1. Let's assume we have a difficulty level = 20 
+2. Now, Let's again assume that we have the previous hash as: 'a5x208fecf9a66be9a2bc7...'
+3. Now we create text as  : text = str(block_number) + transactions + previous_hash + str(nonce)
+4. Now, we pass text to SHA256 Function to generate hash
+5. Finally, that hash prefix must be equal to the number of zeroes as difficulty level.
+6. And Boom!! You have mined successfully.
+
```
Comment (vybhav72954) [inline Python/BitCoin Mining/README.md:26]: You haven't explained anywhere, how to run the script.
Add test cases/screenshots

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
+
+# INPUT : Transactions = '''Neeraj->Zara->50,Nandu->Allu->5''' , difficulty=2 ,
```
Comment (vybhav72954) [inline Python/BitCoin Mining/BitCoin_Mining.py:2]: A test case should reside at the bottom of the file.

**Review by vybhav72954 — CHANGES_REQUESTED** [review]

File: Python/BitCoin Mining/README.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+INPUT : Transactions = '''Neeraj->Zara->50,Nandu->Allu->5''', difficulty=2,
+        Previous_hash = a7sdxa036944e29568d0cff17edbe038f81208fecf9a66be9a2b8321c6ec7.
+OUTPUT : Successfully mined bitcoins with nonce value:336 end mining. Mining took: 12.207852363586426 seconds
+006f74cef9d071afa15c58b38198be14f9b4aabb4cd6f7a44afffd9f6968efcd.
+
+## Draw Backs:
+The script has no drawbacks, but due to the increase in miners, the difficulty level increases, and hence we'd require the best hardware as the fastest wins. 
+
+# Refrences 
+[Here](https://www.investopedia.com/terms/b/bitcoin.asp)
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:41]: Can you use the actual title name of these sites?

File: Python/BitCoin Mining/README.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+[Video](https://www.youtube.com/watch?v=wTC31ZI6QM4)
+
+[Video2](https://www.youtube.com/watch?v=ZhnJ1bkIWWk&t=143s)
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:45]: Can you use the actual title name of these sites?

File: Python/BitCoin Mining/README.md [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+OUTPUT : Successfully mined bitcoins with nonce value:336 end mining. Mining took: 12.207852363586426 seconds
+006f74cef9d071afa15c58b38198be14f9b4aabb4cd6f7a44afffd9f6968efcd.
+
+## Draw Backs:
+The script has no drawbacks, but due to the increase in miners, the difficulty level increases, and hence we'd require the best hardware as the fastest wins. 
+
+# Refrences 
+[Here](https://www.investopedia.com/terms/b/bitcoin.asp)
+
+[Video](https://www.youtube.com/watch?v=wTC31ZI6QM4)
```
Comment (HarshCasper) [inline Python/BitCoin Mining/README.md:43]: Can you use the actual title name of these sites?

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+
+# Function for encoding text to a 64 bit hexadecimal value
+def SHA256(text):
+    """
+    >>> SHA256('Hello')
+    185f8db32271fe25f561a6fc938b2e264306ec304eda518007d1764826381969
+    >>> SHA256('Text')
+    71988c4d8e0803ba4519f0b2864c1331c14a1890bf8694e251379177bfedb5c3
+    """
```
Comment (HarshCasper) [inline Python/BitCoin Mining/BitCoin_Mining.py:15]: Can you use proper docstrings here?

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    return sha256(text.encode("ascii")).hexdigest()
+
+
+# function for guessing nonce value
+def mine(block_number, transactions, previous_hash, prefix_zeros,MAX_NONCE):
```
Comment (HarshCasper) [inline Python/BitCoin Mining/BitCoin_Mining.py:20]: Can you use proper docstrings here?

File: Python/BitCoin Mining/BitCoin_Mining.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+    
+    # Number of prefix zeroes
+    difficulty=int(input('Enter Difficulty level : '))
+
+    # For knowing time taken for mining
+    import time
+    start = time.time()
+    print("start mining")
+
+    previous_hash=input('Enter Previous has value : ')
```
Comment (HarshCasper) [inline Python/BitCoin Mining/BitCoin_Mining.py:60]: It should be `hash` I guess

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: Python/BitCoin Mining/BitCoin_Mining.py (hunk 19 above)
Comment (NEERAJAP2001) [reply Python/BitCoin Mining/BitCoin_Mining.py:60]: @HarshCasper actually the mine function requires ```Previous_hash``` so, I've taken the previous hash as input.
Do you want me to change it to ```hash```?

File: Python/BitCoin Mining/BitCoin_Mining.py (hunk 18 above)
Comment (NEERAJAP2001) [reply Python/BitCoin Mining/BitCoin_Mining.py:20]: I'm sorry I did not understand, can you suggest what could I Improve?
or where did I go wrong?
Thanks

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**